python/pylibcudf/pylibcudf/scalar.pyx

```diff
@@ -152,53 +156,138 @@ cdef Scalar _new_scalar(unique_ptr[scalar] c_obj, DataType dtype):
 
 
 @singledispatch
-def _from_py(py_val):
+def _from_py(py_val, dtype: DataType | None):
     raise TypeError(f"{type(py_val).__name__} cannot be converted to pylibcudf.Scalar")
 
 
 @_from_py.register(dict)
 @_from_py.register(list)
 @_from_py.register(datetime.datetime)
 @_from_py.register(datetime.timedelta)
-def _(py_val):
+def _(py_val, dtype: DataType | None):
     raise NotImplementedError(
         f"Conversion from {type(py_val).__name__} is currently not supported."
     )
 
 
 @_from_py.register(float)
-def _(py_val):
-    cdef DataType dtype = DataType(type_id.FLOAT64)
-    cdef unique_ptr[scalar] c_obj = make_numeric_scalar(dtype.c_obj)
-    (<numeric_scalar[double]*>c_obj.get()).set_value(py_val)
-    cdef Scalar slr = _new_scalar(move(c_obj), dtype)
-    return slr
+def _(py_val: float, dtype: DataType | None):
+    cdef unique_ptr[scalar] c_obj
+    cdef DataType c_dtype
+    if dtype is None:
+        c_dtype = DataType(type_id.FLOAT64)
+    else:
+        c_dtype = <DataType>dtype
+
+    tid = c_dtype.id()
+
+    if tid == type_id.FLOAT32:
+        if abs(py_val) > numeric_limits[float].max():
+            raise OverflowError(f"{py_val} out of range for FLOAT32 scalar")
+        c_obj = make_numeric_scalar(c_dtype.c_obj)
+        (<numeric_scalar[float]*>c_obj.get()).set_value(py_val)
+    elif tid == type_id.FLOAT64:
+        c_obj = make_numeric_scalar(c_dtype.c_obj)
+        (<numeric_scalar[double]*>c_obj.get()).set_value(py_val)
+    else:
+        raise TypeError(f"Cannot convert float to Scalar with dtype {tid.name}")
+
+    return _new_scalar(move(c_obj), dtype)
 
 
 @_from_py.register(int)
-def _(py_val):
-    cdef DataType dtype = DataType(type_id.INT64)
-    cdef unique_ptr[scalar] c_obj = make_numeric_scalar(dtype.c_obj)
-    (<numeric_scalar[int64_t]*>c_obj.get()).set_value(py_val)
-    cdef Scalar slr = _new_scalar(move(c_obj), dtype)
-    return slr
+def _(py_val: int, dtype: DataType | None):
+    cdef unique_ptr[scalar] c_obj
+    cdef DataType c_dtype
+    if dtype is None:
+        c_dtype = DataType(type_id.INT64)
+    else:
+        c_dtype = <DataType>dtype
+    tid = c_dtype.id()
+
+    if tid == type_id.INT8:
+        if abs(py_val) > numeric_limits[int8_t].max():
```

Comment:
Unfortunately this trick does not work for integers as the absolute values of the min and max values are off by 1

```python
In [1]: import numpy as np

In [2]: np.iinfo(np.int8)
Out[2]: iinfo(min=-128, max=127, dtype=int8)
```

Reply:
Oh right, okay will unfold it